### PR TITLE
TEG Refactor

### PR DESCRIPTION
**Pull request by Andromeda-K22**
🆑 
The oil in Engineering's Roboco vendor now has a use. You can lubricate to prevent overloads.
Lubricating the TEGs increases thermal efficiency by 15%. 

Overloads now, functionally, are the same. However, they have some 'scary' aesthetic effects.
When a TEG overloads, it'll make a grumpy message and a noise. There's a chance it'll go a step farther, make a loud BANG, a warning buzzer, shake the camera of anyone in view, and make some smoke.

.... It's perfectly harmless, however!
🆑

### CODE REVIEW

File: code/modules/power/generator.dm [hunk 1]
```diff
@@ -113,12 +125,29 @@
...
+			var/datum/effect/effect/system/spark_spread/s = new /datum/effect/effect/system/spark_spread
+			s.set_up(2, 1, src)
+			s.start()
+			stored_energy *= 0.5
+			if(prob(55))
+				visible_message("<span class='danger'>[src] [pick(soundverb)]!</span>")
+				var/malfsound = pick(soundlist)
+				playsound(src.loc, malfsound, 50, 0, 10)
+				if(prob(20))
+					var/datum/effect/effect/system/smoke_spread/SS = new /datum/effect/effect/system/smoke_spread
```
Comment (scrdest) [inline code/modules/power/generator.dm]: You can just do new() without specifying the type, it's implicit on assignment.

File: code/modules/power/generator.dm [hunk 2]
```diff
@@ -113,12 +125,29 @@
...
-		var/datum/effect/effect/system/spark_spread/s = new /datum/effect/effect/system/spark_spread
-		s.set_up(2, 1, src)
-		s.start()
-		stored_energy *= 0.5
-
+	//exceeding max power is supposed to be bad, have there be a small chance of some visual and audio effects to stress this
+	//remember to lubricate your engines kiddos
+	if(!lubricated)
+		if(effective_gen > max_power && prob(5))
+			var/datum/effect/effect/system/spark_spread/s = new /datum/effect/effect/system/spark_spread
```
Comment (scrdest) [inline code/modules/power/generator.dm]: That's overly verbose and inconsistent with the way you do it for smoke a couple lines below. Speaking of which - you have a /effect/system for sparks called `s`, and one for smoke called `SS` - this is gonna be confusing to maintain.

File: code/modules/power/generator.dm [hunk 3]
```diff
@@ -135,6 +164,13 @@
 		lastgenlev = genlev
 		update_icon()
 	add_avail(effective_gen)
+	if(reagents.has_reagent(/datum/reagent/lube/oil))
+		reagents.remove_any(0.01)
```
Comment (scrdest) [inline code/modules/power/generator.dm]: Subtle bug - you're not guaranteed to actually use up the _oil_ for lubrication if any other reagent winds up in machine.

File: code/modules/power/generator.dm [hunk 4]
```diff
@@ -135,6 +164,13 @@
...
+		thermal_efficiency = 0.80
+		lubricated = 1
+	if(!reagents.has_reagent(/datum/reagent/lube/oil))
```
Comment (scrdest) [inline code/modules/power/generator.dm]: `else` seems like it would make more sense for this.

File: code/modules/power/generator.dm [hunk 5]
```diff
@@ -152,6 +188,15 @@
 		else
 			disconnect_from_network()
 		reconnect()
+	if(istype(W, /obj/item/weapon/reagent_containers))
+		var/obj/item/weapon/reagent_containers/R = W
+		if(!R.reagents.has_reagent(/datum/reagent/lube/oil))
+			audible_message("<span class='warning'>[src] blips in disappointment</span>")
+			playsound(get_turf(src), 'sound/machines/synth_no.ogg', 50, 0)
+			return
+		R.standard_pour_into(user, src)
```
Comment (scrdest) [inline code/modules/power/generator.dm]: You're only checking if the container _has_ oil in it for validation. This means you can put a drop of oil in a bucket of human blood and dump the concoction in the engine just fine. Which is pretty metal. 

Coupled with the other check I commented on, you can lubricate the engine with literally anything from water, through adminordazine, to molten iron.

File: code/modules/power/generator.dm [hunk 6]
```diff
@@ -22,15 +23,26 @@
...
+	create_reagents(120)
 	..()
 	desc = initial(desc) + " Rated for [round(max_power/1000)] kW."
 	spawn(1)
 		reconnect()
 
+/obj/machinery/power/generator/examine(mob/user)
+	..()
+	to_chat(user, "Auxilary tank shows [reagents.total_volume]u of liquid in it.")
+	if(reagents.total_volume == 0)
```
Comment (scrdest) [inline code/modules/power/generator.dm]: `if(lubricated)` would probably be more representative.

File: code/modules/power/generator.dm (hunk 3 above)
Comment (Andromeda-K22) [reply code/modules/power/generator.dm]: There is a check on proc that handles adding the oil.

File: code/modules/power/generator.dm (hunk 5 above)
Comment (Andromeda-K22) [reply code/modules/power/generator.dm]: I can't see a decent way around this, although with the changes on the reagent check in process() i can just make the TEG dump any reagent that isn't oil.

Comment (Naxodile) [reply code/modules/power/generator.dm]: The way to do it is to `reagents.remove_reagent(/datum/reagent/lube/oil, amount)`, however, it'd be pretty odd if somehow only the oil was coming out of something you were pouring.

File: code/modules/power/generator.dm [hunk 7]
```diff
@@ -113,12 +124,29 @@
...
+				playsound(src.loc, malfsound, 50, 0, 10)
+				if(prob(20))
+					var/datum/effect/effect/system/smoke_spread/SM = new()
+					SM.set_up(5, 0, src.loc)
+					playsound(src.loc, 'sound/machines/warning-buzzer.ogg', 50, 1, -3)
+					spawn(2 SECONDS)
+						playsound(src.loc, 'sound/effects/meteorimpact.ogg', 50, 1, -3)
+						for(var/mob/living/M in view(7, src))
+							shake_camera(M, 1, 2)
+						spawn(5)
```
Comment (Naxodile) [inline code/modules/power/generator.dm]: 0.5 SECONDS works just as well. I'm eternally averse to spawn after hearing about its performance, but given that this is nested enough to be classified as a bird's nest I'm sure it'll be fine.

Comment (Andromeda-K22) [reply code/modules/power/generator.dm]: Where do you think the Teshari come from?

**Comment by Glloyd** [comment]
@scrdest Are you happy with this? If so, I'll merge.

**Pull request merged**